geo/src/algorithm/stitch.rs

```diff
@@ -0,0 +1,582 @@
+
+    fn find_outmost_ring<F: GeoFloat>(rings: &[Polygon<F>]) -> Option<usize> {
+        let enumerated_rings = || rings.iter().enumerate();
+        enumerated_rings()
+            .find(|(i, ring)| {
+                enumerated_rings()
+                    .filter(|(j, _)| i != j)
+                    .all(|(_, other)| ring.contains(other))
+            })
+            .map(|(i, _)| i)
+    }
+
+    // if exterior ring exists that contains all other rings, recreate the poly with:
+    //
+    // - exterior ring as exterior
+    // - other rings are counted to interiors
+    // - previously existing interiors are preserved
+    if let Some(outer_index) = find_outmost_ring(&rings) {
+        let exterior = rings.remove(outer_index).exterior().clone();
+        let interiors = poly
+            .interiors()
+            .iter()
+            .cloned()
+            .chain(rings.into_iter().map(|p| p.exterior().clone()))
+            .collect::<Vec<_>>();
+        poly = Polygon::new(exterior, interiors);
+    }
+    poly
+}
+
+/// Inputs to this function is a unordered set of lines that must form a valid multipolygon
+fn stitch_multipolygon_from_lines<F: GeoFloat>(
+    lines: Vec<Line<F>>,
+) -> PolygonStitchingResult<MultiPolygon<F>> {
+    let rings = stitch_rings_from_lines(lines)?;
+
+    fn find_parent_idxs<F: GeoFloat>(
+        ring_idx: usize,
+        ring: &LineString<F>,
+        all_rings: &[LineString<F>],
+    ) -> Vec<usize> {
+        all_rings
+            .iter()
+            .enumerate()
+            .filter(|(other_idx, _)| ring_idx != *other_idx)
+            .filter_map(|(idx, maybe_parent)| {
+                Polygon::new(maybe_parent.clone(), vec![])
+                    .contains(ring)
+                    .then_some(idx)
+            })
+            .collect()
+    }
+
+    // Associates every ring with its parents (the rings that contain it)
+    let parents: HashMap<usize, Vec<usize>> = rings
+        .iter()
+        .enumerate()
+        .map(|(ring_idx, ring)| {
+            let parent_idxs = find_parent_idxs(ring_idx, ring, &rings);
+            (ring_idx, parent_idxs)
+        })
+        .collect();
+
+    // Associates outer rings with their inner rings
+    let mut polygons_idxs: HashMap<usize, Vec<usize>> = HashMap::default();
+
+    // For each ring, we check how many parents it has  otherwise it's an outer ring
+    //
+    // This is important in the scenarios of "donuts" where you have an outer donut shaped
+    // polygon which completely contains a smaller polygon inside its hole
+    for (ring_index, parent_idxs) in parents.iter() {
+        let parent_count = parent_idxs.len();
+
+        // if it has an even number of parents, it's an outer ring so we can just add it if it's
+        // missing
+        if parent_count % 2 == 0 {
+            polygons_idxs.entry(*ring_index).or_default();
+            continue;
+        }
+
+        // the direct parent is the parent which has itself the most parents
```

Comment:
To make sure I'm understanding, is a "direct parent" also the "closest" parent?

e.g. Given that `A` contains `B` contains `C`

- `A` is a parent of `B` and `C`.
- `C`'s parents are `A` and `B`.
- `C`'s direct parent is `B` - there can be only one, right?

Reply:
Exactly 👍🏼